proxy.conf.template

```diff
@@ -15,19 +16,34 @@ server {
     # be idendified by either domain
     server_name ${PROXY_DOMAIN} ${PROXY_ORIGIN};
     root /var/www/html;
-
     listen ${PORT};
 
+    # we need a resolver configured so we can look up domains when redirecting
+    resolver ${RESOLVER};
```

Comment:
with my worrying hat on, prodding some assumptions - are we talking to the resolver in a secure way? or assuming that because we're using `127.0.0.1` it's secure? in which case, why parameterise it in the first place & open the possibility for an insecure configuration?

Reply:
This is a quirk of nginx design.

By default, it resolves all addresses found in the config files at start up using the system resolver. When it runs, it doesn't actually have DNS look ups available unless you tell it to use a resolver. As we don't control the domain returned in the redirect from ghcr.io, we need to explicitly enable a resolver.

Nginx doesn't provide a way to use the system resolver, you have to be explicit. I would have defaulted to 127.0.0.1, but dokku2 is on debian, and the host doesn't have a local DNS apparently - it's got DO DNS ips configured.

We could work round this by adding the path of the Location: header to hardcoded domain (currently github-production-container-registry.s3.amazonaws.com). This would remove the need for resolver config. But it would break if ghcr.io ever started sending different domains. And given the definitely *didn't* redirect ~5 weeks ago when I first tested this, but they do know, it might change.

Reply:
I think that info definitely lives in the documentation somewhere

Reply:
yeah, although I might be tempted to have:
docker-proxy repo docs
  * public
  * resolver (e.g. `127.0.0.1`) goes here
sysadmin repo docs
  * private
  * we deploy on DO, the resolver setting is `DO's DNS`, etc
wdyt?

Reply:
So, if we hardcoded the domain, it would use the system resolver, which would be DO's DNS servers? So if someone was trying to put a proxy between us & DO's DNS servers, hardcoding it would have the same potential vuln?

Reply:
Oh, also, hardcoding it would presumably only do the look up once - at nginx startup - so we'd have to reload/restart nginx every time that went away (rather than on some timeout as with a resolver). That would suck...

Reply:
Just feels like - we're doing all this work to get a secure path for the docker images, is annoying to lean on DNS lookups potentially sent in the clear.

Reply:
Yes, hardcoding doesn't protect us from DNS poisoning. It just a) avoids the need for explicit resolver and b) acts as a domain whitelist. But yes we'd have to restart if ghcr.io change their domains or IPs.

I mean securing DNS for dokku2 may make sense (i.e. whitelisting dns lookups), but I think it's orthogonal to this change. We already trust it for resolving ghcr.io itself. And everything is https, so DNS poisoning alone is not sufficient to subvert. Additionally, the main benefit is preventing docke push from the backend, which still works even if DNS is compromised.

What's the attack scenario you envisage with poisoned DNS?

Reply:
Yeah, I hadn't taken into account that the S3 urls would guaranteed https.